client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -1,182 +1,136 @@
-              createConnection(key, callback)
+  /**
+    * A Circular Buffer That Does Not Block on Enqueue but,
+    * will dequeue and shutdown the oldest connection in the buffer.
+    */
+  private def connectionBuffer(implicit async: Async[Task]): Task[BuildDestroyQueue[Task, RequestKey, A]] =
+    Semaphore.apply(maxTotal.toLong).flatMap { permits =>
+    unbounded[Task,A].map { q =>
+      new BuildDestroyQueue[Task,RequestKey, A] {
+        def upperBound: Option[Int] = Some(maxTotal)
+        def enqueue1(a:A): Task[Unit] =
+          permits.tryDecrement.flatMap { b => if (b) q.enqueue1(a) else q.dequeue1.flatMap(destroy) >> q.enqueue1(a) }
+        def offer1(a: A): Task[Boolean] =
+          enqueue1(a).as(true)
+        def dequeue1: Task[A] = cancellableDequeue1.flatMap { _._1 }
+        def dequeueBatch1(batchSize: Int): Task[NonEmptyChunk[A]] = cancellableDequeueBatch1(batchSize).flatMap { _._1 }
+        def cancellableDequeue1: Task[(Task[A], Task[Unit])] = cancellableDequeueBatch1(1).map { case (deq,cancel) => (deq.map(_.head),cancel) }
+        def cancellableDequeueBatch1(batchSize: Int): Task[(Task[NonEmptyChunk[A]],Task[Unit])] =
+          q.cancellableDequeueBatch1(batchSize).map { case (deq,cancel) => (deq.flatMap(a => permits.incrementBy(a.size.toLong).as(a)), cancel) }
+
+        def size = q.size
+        def full: Signal[Task, Boolean] = q.size.map(_ >= maxTotal)
+        def available: Signal[Task, Int] = q.size.map(maxTotal - _)
+
+        override def build(key: RequestKey): Task[Unit] =
+          permits.tryDecrement.flatMap { b =>
+            if (b) builder(key).flatMap(connection => enqueue1(connection))
+            else q.dequeue1.flatMap(destroy) >> builder(key).flatMap(connection => enqueue1(connection))
           }
-        }
-        else {
-          allocated -= 1
 
-          if (!connection.isClosed) {
-            logger.debug(s"Connection returned was busy.  Shutting down: ${stats}")
-            connection.shutdown()
-          }
+        override def destroy(connection: A): Task[Unit] =
+          connection.isClosed.flatMap{ closed => if (!closed) connection.shutdown() else Task.now(()) }
 
-          if (waitQueue.nonEmpty) {
-            logger.debug(s"Connection returned could not be recycled, new connection needed: ${stats}")
-            val Waiting(key, callback) = waitQueue.dequeue()
-            createConnection(key, callback)
+      }
+    }}
+
+  /**
+    * This method is similar to dequeue except that it works with the key for a Request rather than an element of
+    * the queue directly.
+    * If the buffer is full, we dequeue an element and if it is still open, we either use it if it matches the
+    * RequestKey or we add it back to the queue. If the connection is closed, we add a connection to the pool with the
+    * new RequestKey. If the buffer is not full we create a connection and add it to the queue. This builds the buffer
+    * pool initially.
```

Comment:
What is as needed in this case? Like when I get a request for a connection, then I search through N idle connections, if none match RequestKey then build? With the old mutable queue, you were using lookup, which is not a feature in fs2s Queue.